File: stacks/opinionated/rails-app/app/controllers/orders_controller.rb
class OrdersController < ApplicationController
  VALID_ORDER_STATUSES = %w[created paid shipped cancelled].freeze
  ALLOWED_STATUS_TRANSITIONS = {
    "created" => %w[paid cancelled],
    "paid" => %w[shipped cancelled],
    "shipped" => [],
    "cancelled" => []
  }.freeze

  def create
    payload = params.permit(:user_id, items: [:product_id, :quantity])
    user_id = require_user_id(payload)
    items = validate_items(payload[:items])
    user = User.find(user_id)
    order = Order.create_with_items!(user: user, items: items)

    render json: order.as_api_json, status: :created
  end

  def index
    orders = Order.detailed.order(:id)
    status_filter = params[:status]
    user_id_filter = params[:user_id]

    if status_filter.present?
      return render_error("Query parameter 'status' is invalid", :unprocessable_entity) unless VALID_ORDER_STATUSES.include?(status_filter)

      orders = orders.where(status: status_filter)
    end

    if user_id_filter.present?
      return render_error("Query parameter 'user_id' is invalid", :unprocessable_entity) unless user_id_filter.to_s.match?(/\A\d+\z/)

      orders = orders.where(user_id: user_id_filter.to_i)
    end

    render json: orders.map(&:as_api_json)
  end

  def show
    render json: Order.fetch_detailed!(params[:id]).as_api_json
  end

  def status
    order = Order.find(params[:id])
    status_value = params.permit(:status)[:status]

    return render_error("Field 'status' is required", :unprocessable_entity) unless status_value.is_a?(String) && status_value.strip != ""
    return render_error("Field 'status' is invalid", :unprocessable_entity) unless VALID_ORDER_STATUSES.include?(status_value)
    return render_error("Invalid order status transition", :conflict) unless ALLOWED_STATUS_TRANSITIONS.fetch(order.status).include?(status_value)

    order.update!(status: status_value)
    render json: Order.fetch_detailed!(order.id).as_api_json
  end

  private

  def require_user_id(payload)
    user_id = payload[:user_id]
    raise ActionController::BadRequest, "Field 'user_id' is required" unless user_id.is_a?(Integer)

    user_id
  end

  def validate_items(items)
    raise ActionController::BadRequest, "Order must contain at least one item" unless items.is_a?(Array) && items.any?

    items.map do |item|
      product_id = item[:product_id]
      quantity = item[:quantity]

      raise ActionController::BadRequest, "Field 'product_id' is required" unless product_id.is_a?(Integer)
      raise ActionController::BadRequest, "Field 'quantity' must be greater than zero" unless quantity.is_a?(Integer) && quantity.positive?

      { product_id: product_id, quantity: quantity }
    end
  end
end

File: stacks/opinionated/laravel-app/app/Http/Controllers/OrdersController.php
<?php

namespace App\Http\Controllers;

use App\Models\Order;
use App\Models\OrderItem;
use App\Models\Product;
use App\Models\User;
use Illuminate\Http\JsonResponse;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\DB;
use Illuminate\Support\Facades\Validator;

class OrdersController extends Controller
{
    private const VALID_STATUSES = ['created', 'paid', 'shipped', 'cancelled'];
    private const ALLOWED_STATUS_TRANSITIONS = [
        'created' => ['paid', 'cancelled'],
        'paid' => ['shipped', 'cancelled'],
        'shipped' => [],
        'cancelled' => [],
    ];

    public function store(Request $request): JsonResponse
    {
        $validator = Validator::make($request->all(), [
            'user_id' => ['required', 'integer'],
            'items' => ['required', 'array', 'min:1'],
            'items.*.product_id' => ['required', 'integer'],
            'items.*.quantity' => ['required', 'integer', 'min:1'],
        ]);

        if ($validator->fails()) {
            return response()->json(['detail' => $validator->errors()->first()], 422);
        }

        $validated = $validator->validated();
        $user = User::query()->find($validated['user_id']);

        if (! $user) {
            return response()->json(['detail' => 'User not found'], 404);
        }

        $productIds = collect($validated['items'])->pluck('product_id')->unique()->values();
        $products = Product::query()->whereIn('id', $productIds)->get()->keyBy('id');

        if ($products->count() !== $productIds->count()) {
            return response()->json(['detail' => 'Product not found'], 404);
        }

        $order = DB::transaction(function () use ($validated, $user, $products) {
            $total = collect($validated['items'])->reduce(function ($sum, $item) use ($products) {
                $product = $products->get($item['product_id']);

                return $sum + ($product->price * $item['quantity']);
            }, 0.0);

            $order = Order::query()->create([
                'user_id' => $user->id,
                'total' => $total,
                'status' => 'created',
            ]);

            foreach ($validated['items'] as $item) {
                $product = $products->get($item['product_id']);

                $order->items()->create([
                    'product_id' => $product->id,
                    'quantity' => $item['quantity'],
                    'unit_price' => $product->price,
                ]);
            }

            return $this->findDetailedOrder($order->id);
        });

        return response()->json($this->serializeOrder($order), 201);
    }

    public function index(Request $request): JsonResponse
    {
        $query = Order::query()
            ->with(['user:id,name', 'items.product:id,name'])
            ->orderBy('id');

        if ($request->has('status')) {
            if (! in_array($request->input('status'), self::VALID_STATUSES, true)) {
                return response()->json(['detail' => "Query parameter 'status' is invalid"], 422);
            }
            $query->where('status', $request->input('status'));
        }

        if ($request->has('user_id')) {
            if (! ctype_digit((string) $request->input('user_id')) || (int) $request->input('user_id') <= 0) {
                return response()->json(['detail' => "Query parameter 'user_id' is invalid"], 422);
            }
            $query->where('user_id', (int) $request->input('user_id'));
        }

        $orders = $query->get();

        return response()->json($orders->map(fn (Order $order) => $this->serializeOrder($order)));
    }

    public function show(int $id): JsonResponse
    {
        $order = $this->findDetailedOrder($id);

        if (! $order) {
            return response()->json(['detail' => 'Order not found'], 404);
        }

        return response()->json($this->serializeOrder($order));
    }

    public function updateStatus(Request $request, int $id): JsonResponse
    {
        $validator = Validator::make($request->all(), [
            'status' => ['required', 'string'],
        ]);

        if ($validator->fails()) {
            return response()->json(['detail' => $validator->errors()->first()], 422);
        }

        $order = Order::query()->find($id);

        if (! $order) {
            return response()->json(['detail' => 'Order not found'], 404);
        }

        $nextStatus = $validator->validated()['status'];
        if (! in_array($nextStatus, self::VALID_STATUSES, true)) {
            return response()->json(['detail' => 'Field status is invalid'], 422);
        }
        if (! in_array($nextStatus, self::ALLOWED_STATUS_TRANSITIONS[$order->status] ?? [], true)) {
            return response()->json(['detail' => 'Invalid order status transition'], 409);
        }

        $order->update([
            'status' => $nextStatus,
        ]);

        return response()->json($this->serializeOrder($this->findDetailedOrder($order->id)));
    }

    private function findDetailedOrder(int $id): ?Order
    {
        return Order::query()
            ->with(['user:id,name', 'items.product:id,name'])
            ->find($id);
    }

    private function serializeOrder(Order $order): array
    {
        return [
            'id' => $order->id,
            'user' => [
                'id' => $order->user->id,
                'name' => $order->user->name,
            ],
            'items' => $order->items
                ->sortBy('id')
                ->map(fn (OrderItem $item) => [
                    'product_id' => $item->product_id,
                    'product_name' => $item->product->name,
                    'quantity' => $item->quantity,
                    'unit_price' => (float) $item->unit_price,
                ])
                ->values()
                ->all(),
            'item_count' => $order->items->count(),
            'total' => (float) $order->total,
            'status' => $order->status,
            'created_at' => $order->created_at->toISOString(),
        ];
    }
}
